src/join-9.sol

```diff
@@ -35,65 +33,99 @@ interface GemLike {
 
-contract GemJoin9 is LibNote {
-    // --- Auth ---
-    mapping (address => uint256) public wards;
-    function rely(address usr) external note auth { wards[usr] = 1; }
-    function deny(address usr) external note auth { wards[usr] = 0; }
-    modifier auth {
-        require(wards[msg.sender] == 1, "GemJoin/not-authorized");
-        _;
-    }
+contract GemJoin9 {
+    // --- Data ---
+    mapping (address => uint256) public wards; // Auth
 
-    VatLike public vat;   // CDP Engine
-    bytes32 public ilk;   // Collateral Type
-    GemLike public gem;
-    uint256 public dec;
-    uint256 public live;  // Active Flag
+    uint256 public live;                       // Active Flag
     uint256 public total;
 
+    VatLike public immutable vat;              // CDP Engine
+    bytes32 public immutable ilk;              // Collateral Type
+    GemLike public immutable gem;
+    uint256 public immutable dec;
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event Join(address indexed usr, uint256 wad);
+    event Exit(address indexed usr, uint256 wad);
+    event Cage();
+
     constructor(address vat_, bytes32 ilk_, address gem_) public {
-        wards[msg.sender] = 1;
-        live = 1;
         vat = VatLike(vat_);
         ilk = ilk_;
         gem = GemLike(gem_);
-        dec = gem.decimals();
+        dec = GemLike(gem_).decimals();
+        live = 1;
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    // --- Auth ---
+    modifier auth {
+        require(wards[msg.sender] == 1, "GemJoin9/not-authorized");
+        _;
+    }
+
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
     }
-    function cage() external note auth {
+
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    function cage() external auth {
         live = 0;
+        emit Cage();
     }
+
+    // --- Math ---
     function add(uint256 x, uint256 y) internal pure returns (uint256 z) {
-        require((z = x + y) >= x);
+        require((z = x + y) >= x, "GemJoin9/overflow");
     }
     function sub(uint256 x, uint256 y) internal pure returns (uint256 z) {
-        require((z = x - y) <= x);
+        require((z = x - y) <= x, "GemJoin9/underflow");
     }
+
     // Allow dss-proxy-actions to send the gems with only 1 transfer
     // This should be called via token.transfer() followed by gemJoin.join() atomically or
     // someone else can steal your tokens
-    function join(address usr) public note returns (uint256) {
-        require(live == 1, "GemJoin/not-live");
+    function join(address usr) external returns (uint256 wad) {
+        wad = _join(usr);
+
+        emit Join(usr, wad);
+    }
+
+    function join(address usr, uint256 wad) external {
+        require(gem.transferFrom(msg.sender, address(this), wad), "GemJoin9/failed-transfer");
+
+        _join(usr);
```

Comment:
I might re-write this function a little to match the other join.

this join should return `uint256 wad`, and maybe call the input param `amt`

`transferFrom()` the `amt`, and set line 106 to `wad = _join(usr);` since there may be some amount of gem in this contract already that gets joined in addition to what's transferred in.